### ISSUE TITLE
corner pooling and `torch.cummax`

### ISSUE THREAD

**Issue by Scalsol** [open-mmlab/mmdetection#3008]
Hi, I notice that you are adding corner pooling module as a new feature. But in PyTorch 1.5, a new function named `torch.cummax` is introduced, which I think behaves just like corner pooling. As a known fact that corner pooling runs at a slow speed, would `torch.cummax` runs faster?

**Comment by hellock** [comment]
Thanks for the reference. @tianyuandu may have a look.

**Comment by tianyuandu** [comment]
Hi @Scalsol , thanks for your suggestion. I found the output of `torch.cummax` and our corner pool is same, but the grad is mismatching.

Here is the verification script:
```python
import torch

from mmdet.ops import CornerPool


if __name__ == '__main__':
    input_tensor = torch.tensor([[[[0, 0, 0, 0, 0], [2, 1, 3, 0, 2], 
                                   [5, 4, 1, 1, 6], [0, 0, 0, 0, 0], 
                                   [0, 0, 0, 0, 0]]]]).float().cuda()
    cum_tensor = input_tensor.detach().requires_grad_()
    pool_tensor = input_tensor.detach().requires_grad_()

    pool = CornerPool('bottom')
    pool_out = pool(pool_tensor)
    pool_loss = pool_out.sum()
    pool_loss.backward()

    cum_out = torch.cummax(cum_tensor, dim=2)[0]
    cum_loss = cum_out.sum()
    cum_loss.backward()

    print("pool_out:")
    print(pool_out)
    print("cum_out:")
    print(cum_out)
    print("pool_tensor's grad:")
    print(pool_tensor.grad)
    print("cum_tensor's grad:")
    print(cum_tensor.grad)
```

and the output is:
```python
pool_out:
tensor([[[[0., 0., 0., 0., 0.],
          [2., 1., 3., 0., 2.],
          [5., 4., 3., 1., 6.],
          [5., 4., 3., 1., 6.],
          [5., 4., 3., 1., 6.]]]], device='cuda:0',
       grad_fn=<BottomPoolFunctionBackward>)
cum_out:
tensor([[[[0., 0., 0., 0., 0.],
          [2., 1., 3., 0., 2.],
          [5., 4., 3., 1., 6.],
          [5., 4., 3., 1., 6.],
          [5., 4., 3., 1., 6.]]]], device='cuda:0', grad_fn=<CummaxBackward>)
pool_tensor's grad:
tensor([[[[1., 1., 1., 2., 1.],
          [1., 1., 4., 0., 1.],
          [3., 3., 0., 3., 3.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')
cum_tensor's grad:
tensor([[[[1., 1., 1., 1., 1.],
          [1., 1., 4., 1., 1.],
          [3., 3., 0., 3., 3.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')
```

Do you have any opinion about this?

**Comment by Scalsol** [comment]
@tianyuandu It seems that the gradient difference occurs when two elements have the same value. When the current value is the same as the max value, CornerPool prefers to use the original index, while cummax would update the index. I think this assumption could be easily verified. In my opinion both implementation are current.

BTW, as the values of different position in a feature map could hardly be the same when training the network, I think this gradient difference would not influence the final result.

And one more thing, will torch.cummax runs faster?

**Comment by tianyuandu** [comment]
> @tianyuandu It seems that the gradient difference occurs when two elements have the same value. When the current value is the same as the max value, CornerPool prefers to use the original index, while cummax would update the index. I think this assumption could be easily verified. In my opinion both implementations are correct.
> 
> BTW, as the values of different position in a feature map could hardly be the same when training the network, I think this gradient difference would not influence the final result.
> 
> And one more thing, will torch.cummax runs faster?

I think your assumption is correct. I expand the input tensor to
```python
tensor([[[[0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0],
          [2, 1, 3, 0, 2],
          [2, 1, 3, 0, 2],
          [5, 4, 1, 1, 6],
          [0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0]]]]
```
and the grad of two ops are:
```python
corner_pool:
tensor([[[[2., 2., 2., 4., 2.],
          [0., 0., 0., 0., 0.],
          [2., 2., 5., 0., 2.],
          [0., 0., 0., 0., 0.],
          [3., 3., 0., 3., 3.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')

cummax:
tensor([[[[1., 1., 1., 1., 1.],
          [1., 1., 1., 1., 1.],
          [1., 1., 1., 1., 1.],
          [1., 1., 4., 1., 1.],
          [3., 3., 0., 3., 3.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')
```
I'm also checking whether using different op would effect on final mAP, I will reply once I get the result. And using `torch.cummax` could get ~10% speedup on V100(from 1.86s/iter to 1.68s/iter, img_per_gpu=6).

**Comment by tianyuandu** [comment]
> @tianyuandu It seems that the gradient difference occurs when two elements have the same value. When the current value is the same as the max value, CornerPool prefers to use the original index, while cummax would update the index. I think this assumption could be easily verified. In my opinion both implementations are correct.
> 
> BTW, as the values of different position in a feature map could hardly be the same when training the network, I think this gradient difference would not influence the final result.
> 
> And one more thing, will torch.cummax runs faster?

I think your assumption is correct. I expand the input tensor to
```python
tensor([[[[0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0],
          [2, 1, 3, 0, 2],
          [2, 1, 3, 0, 2],
          [5, 4, 1, 1, 6],
          [0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0]]]]
```
and the grad of two ops are:
```python
corner_pool:
tensor([[[[2., 2., 2., 4., 2.],
          [0., 0., 0., 0., 0.],
          [2., 2., 5., 0., 2.],
          [0., 0., 0., 0., 0.],
          [3., 3., 0., 3., 3.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')

cummax:
tensor([[[[1., 1., 1., 1., 1.],
          [1., 1., 1., 1., 1.],
          [1., 1., 1., 1., 1.],
          [1., 1., 4., 1., 1.],
          [3., 3., 0., 3., 3.],
          [0., 0., 0., 0., 0.],
          [0., 0., 0., 0., 0.]]]], device='cuda:0')
```
I'm also checking whether using different op would effect on final mAP, I will reply once I get the result. And using `torch.cummax` could get ~10% speedup on V100(from 1.86s/iter to 1.68s/iter, img_per_gpu=6).

**Comment by tianyuandu** [comment]
@Scalsol  The verification experiment is done. Both `torch.cummax` and our corner_pool could reach 41.0 mAP with `batch_size=6 * 16` on V100. Now I'm going to open a PR to fix this problem. Thanks for your suggestion!

**Comment by Scalsol** [comment]
@tianyuandu glad to hear that!
